tests/rust/pipe-test/async_close/mod.rs

```diff
@@ -0,0 +1,220 @@
+        Err(e) if e.errno == libc::ETIMEDOUT => {
+            anyhow::bail!("wait() async_close() should not timeout")
+        },
+        Ok(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+        Err(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+    }
+
+    // Fail to close the pipe again.
+    match libos.async_close(pipeqd) {
+        Err(e) if e.errno == libc::EBADF => Ok(()),
+        Ok(_) => anyhow::bail!("async_close() a pipe multiple times should fail"),
+        Err(e) => anyhow::bail!("async_close() failed ({})", e),
+    }
+}
+
+/// Attempts to asynchronously close the same pipe multiple times.
+fn async_close_pipe_multiple_times_2(libos: &mut LibOS, pipe_name: &str) -> Result<()> {
+    let mut cancelled: bool = false;
+    let mut closed: bool = false;
+
+    // Create a pipe.
+    let pipeqd: QDesc = match libos.create_pipe(pipe_name) {
+        Ok(pipeqd) => pipeqd,
+        Err(e) => anyhow::bail!("create_pipe() failed ({})", e),
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt1: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt2: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt1, Some(Duration::from_micros(0))) {
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_CLOSE && qr.qr_ret == 0 => closed = true,
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED && qr.qr_ret == libc::ECANCELED => cancelled = true,
+        Ok(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+        Err(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+    }
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt2, Some(Duration::from_micros(0))) {
+        Ok(qr) => {
+            if cancelled && qr.qr_opcode == demi_opcode_t::DEMI_OPC_CLOSE && qr.qr_ret == 0 {
+                return Ok(());
+            }
+            if closed && qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED && qr.qr_ret == libc::EBADF {
+                return Ok(());
+            }
```

Comment:
can you change this to two match arms to match the others?

Reply:
I can't because I am testing for boolean conditions as well.